servers/src/main/java/tachyon/master/journal/JournalWriter.java

```diff
@@ -324,16 +329,22 @@ public synchronized void flush() throws IOException {
       if (mIsClosed) {
         return;
       }
-      mOutputStream.flush();
-      if (mOutputStream instanceof FSDataOutputStream) {
-        ((FSDataOutputStream) mOutputStream).sync();
+      mDataOutputStream.flush();
+      if (mRawOutputStream instanceof FSDataOutputStream) {
+        // The output stream directly created by {@link UnderFileSystem} may be
+        // {@link FSDataOutputStream}, which means the under filesystem is HDFS, but
+        // {@link DataOutputStream#flush} won't flush the data to HDFS, so we need to call
+        // {@link FSDataOutputStream#sync} to actually flush data to HDFS.
+        ((FSDataOutputStream) mRawOutputStream).sync();
```

Comment:
Is this part the problem, since `mOutputStream` would not be a `FSDataOutputStream`?